### ISSUE TITLE
Python error still Present?

### ISSUE THREAD

**Issue by styphonthal** [m50/ha-fallback-conversation#17]
I updated to the 1.0.3 version, but even with this I still get the same error message, it is with using Home assistant 1st, and openai 2nd. 

Logger: homeassistant.components.assist_pipeline.pipeline
Source: components/assist_pipeline/pipeline.py:994
integration: Assist pipeline ([documentation](https://www.home-assistant.io/integrations/assist_pipeline), [issues](https://github.com/home-assistant/core/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+assist_pipeline%22))
First occurred: 8:19:04 PM (1 occurrences)
Last logged: 8:19:04 PM

Unexpected error during intent recognition
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/assist_pipeline/pipeline.py", line 994, in recognize_intent
    conversation_result = await conversation.async_converse(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/conversation/agent_manager.py", line 108, in async_converse
    result = await method(conversation_input)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fallback_conversation/__init__.py", line 90, in async_process
    agent_names[agent_id],
    ~~~~~~~~~~~^^^^^^^^^^
KeyError: 'ec096d1807e1cc6a84562ffa2c44f4db'

**Comment by m50** [comment]
Hi @styphonthal, can you recreate this consistently? What are you asking? What is the exact steps you are taking to recreate it? Could you enable debug logging for fallback agent and provide the full logs?

I am unable to recreate this issue (I was prior to 1.0.3 though), so if I can have a way to recreate it I can look into it.

**Comment by styphonthal** [comment]
I am using the latest version of Home assistant OS, and esphome. I am using a ESP32-Box-s3.  Using fallback with HA as primary, Openai as secondary, I can recreate this error every time by saying a command I know will fail HA, but that Openai can do. I know this is Fallbacks problem as I can ask the exact same question (checking whisper logs to confirm) and if I replace Fallback with Openai it works. I can do commands that will go to HA without problem. This seems to be a problem caused by the latest updates, as previously it worked without problem. Turning on debugging gives me the same stuff. Of note, it still says "1.0.0" in the Integrations area, but I don't think this changes with updating to 1.0.3. I have deleted both the integration and in HACs, then restarted, then reinstalled with no change. 

Logger: homeassistant.components.assist_pipeline.pipeline
Source: components/assist_pipeline/pipeline.py:994
integration: Assist pipeline (documentation, issues)
First occurred: 10:24:01 AM (2 occurrences)
Last logged: 10:25:09 AM

Unexpected error during intent recognition
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/assist_pipeline/pipeline.py", line 994, in recognize_intent
    conversation_result = await conversation.async_converse(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/conversation/agent_manager.py", line 108, in async_converse
    result = await method(conversation_input)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fallback_conversation/__init__.py", line 90, in async_process
    agent_names[agent_id],
    ~~~~~~~~~~~^^^^^^^^^^
KeyError: 'ec096d1807e1cc6a84562ffa2c44f4db'

**Comment by m50** [comment]
You didn't provide any of the debug logging. Could you please enable debug logging and provide me the full logs leading up to it?

When you reinstalled, did you delete and set back up the fallback conversation agent? I wonder if maybe something is saved that has it misconfigured?

> it still says "1.0.0" in the Integrations area,

Good to know, I'll look into fixing that in 1.0.4

**Comment by m50** [comment]
I just released v1.0.4, which *may* have a fix for it, but no guarantees. The main thing I was fixing was #12, but I threw something in there that may "fix" (more like, workaround) this. Though since I cannot recreate it myself, no guarantees.

**Comment by styphonthal** [comment]
With debug logging on, the error is the same:

2024-06-30 19:25:45.576 ERROR (MainThread) [homeassistant.components.assist_pipeline.pipeline] Unexpected error during intent recognition
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/assist_pipeline/pipeline.py", line 994, in recognize_intent
    conversation_result = await conversation.async_converse(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/conversation/agent_manager.py", line 108, in async_converse
    result = await method(conversation_input)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fallback_conversation/__init__.py", line 91, in async_process
    agent_names[agent_id] or "[unknown]",
    ~~~~~~~~~~~^^^^^^^^^^
KeyError: 'ec096d1807e1cc6a84562ffa2c44f4db'
2024-06-30 19:27:20.775 ERROR (MainThread) [homeassistant.components.assist_pipeline.pipeline] Unexpected error during intent recognition
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/assist_pipeline/pipeline.py", line 994, in recognize_intent
    conversation_result = await conversation.async_converse(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/conversation/agent_manager.py", line 108, in async_converse
    result = await method(conversation_input)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fallback_conversation/__init__.py", line 91, in async_process
    agent_names[agent_id] or "[unknown]",
    ~~~~~~~~~~~^^^^^^^^^^
KeyError: 'ec096d1807e1cc6a84562ffa2c44f4db'

**Comment by styphonthal** [comment]
So i removed it from integrations and from HACS. I manually went and deleted the directory and files also. Same error.
So i tried using the "assist" button to type in my response to see if anything changes, same error. Yet it works if I change it to openai. If I ask a question that "primary" can answer (my case home assistant), it will not give this error and will work, from this debug log:

2024-06-30 19:51:56.192 ERROR (MainThread) [homeassistant.components.assist_pipeline.pipeline] Unexpected error during intent recognition
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/assist_pipeline/pipeline.py", line 994, in recognize_intent
    conversation_result = await conversation.async_converse(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/conversation/agent_manager.py", line 108, in async_converse
    result = await method(conversation_input)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fallback_conversation/__init__.py", line 91, in async_process
    agent_names[agent_id] or "[unknown]",
    ~~~~~~~~~~~^^^^^^^^^^
KeyError: 'ec096d1807e1cc6a84562ffa2c44f4db'
2024-06-30 19:54:47.786 DEBUG (MainThread) [custom_components.fallback_conversation] Processing in en using conversation.home_assistant with debug level low: check if any doors are open
2024-06-30 19:54:48.056 ERROR (MainThread) [homeassistant.components.assist_pipeline.pipeline] Unexpected error during intent recognition
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/assist_pipeline/pipeline.py", line 994, in recognize_intent
    conversation_result = await conversation.async_converse(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/conversation/agent_manager.py", line 108, in async_converse
    result = await method(conversation_input)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fallback_conversation/__init__.py", line 91, in async_process
    agent_names[agent_id] or "[unknown]",
    ~~~~~~~~~~~^^^^^^^^^^
KeyError: 'ec096d1807e1cc6a84562ffa2c44f4db'

**Comment by styphonthal** [comment]
So, I was using openAi extended conversation, and decided to try to see if I can get the same error with the integrated OpenAi/Chatgpt, which I do not. 

What is interesting is that the openAi extended conversation worked with Fallback conversation until the recent HA updates

**Comment by mikey60** [comment]
First, thanks for a great integration!  After upgrading from version 1.0.2 to 1.0.4 I get the same error as styphonthal except with a different KeyError value that I believe is the ID of the fallback conversation agent I am using.   I am using HA version 2024.6.1.  I am running the custom "Ollama Conversation" integration that points to a local server running Ollama version 1.48 with a llama3 model as the Fallback Conversation Agent.  I reverted back to version 1.0.2 and it is working properly.

**Comment by styphonthal** [comment]
> First, thanks for a great integration! After upgrading from version 1.02 to 1.04 I get the same error as styphonthal except with a different KeyError value that I believe is the ID of the fallback conversation agent I am using. I am using HA version 2024.6.1. I am running the custom "Ollama Conversation" integration that points to a local server running Ollama version 1.48 with a llama3 model as the Fallback Conversation Agent. I reverted back to version 1.02 and it is working properly.

Thank you, rolling back to 1.02 allows me to use "openai extended" which is much better than HAs openai integration

**Comment by Rafaille** [comment]
First, thank you so much @m50 for developing such a useful component.
I have the same error here with Fallback v1.0.4 and HA up to date.
Do you have all the information you need to look into this? If not, please let me know what you need and I will provide it.

**Comment by m50** [comment]
> So, I was using openAi extended conversation, and decided to try to see if I can get the same error with the integrated OpenAi/Chatgpt, which I do not.
> 
> What is interesting is that the openAi extended conversation worked with Fallback conversation until the recent HA updates

>Thank you, rolling back to 1.02 allows me to use "openai extended" which is much better than HAs openai integration

So wait, sorry, it's not clear here, but is the Extended conversation agent (from HACS) not working now, but *does* work in 1.0.2? That's odd...

This would be way easier to test if I could recreate it.

I am curious @styphonthal or @Rafaille, if you delete and recreate your OpenAI integration, does it then work? I am curious what is different between my setup and yours. The one big difference is I setup my OpenAI integration in the latest version of HomeAssistant.

If I create a debug build that provides me additional information, could one of you provide me with a debug log from the debug build?

**Comment by m50** [comment]
> First, thanks for a great integration! After upgrading from version 1.0.2 to 1.0.4 I get the same error as styphonthal except with a different KeyError value that I believe is the ID of the fallback conversation agent I am using. I am using HA version 2024.6.1. I am running the custom "Ollama Conversation" integration that points to a local server running Ollama version 1.48 with a llama3 model as the Fallback Conversation Agent. I reverted back to version 1.0.2 and it is working properly.

Reverting back fixed it? Interesting, considering 1.0.3 had actual fixes for this issue for some people. Super weird that it worked for some and broke things for others. I'll have to take a closer look at what @ov1d1u did, maybe I can figure out what the issue is.

**Comment by Rafaille** [comment]
> I am curious @styphonthal or @Rafaille, if you delete and recreate your OpenAI integration, does it then work? I am curious what is different between my setup and yours. The one big difference is I setup my OpenAI integration in the latest version of HomeAssistant.

I have just completely uninstalled the Extended OpenAI integration and component, then redownloaded and reinstalled it and I still see exactly the same error.

> If I create a debug build that provides me additional information, could one of you provide me with a debug log from the debug build?

I am happy to do that yes. I will be away until Monday but I can definitely give it a go when I come back.
Thanks for the support!

**Comment by m50** [comment]
Ok, good news (I guess?), I am now having the same problem with Extended OpenAI Integration, so I am able to dig into it myself. I will let you know if I need any assistance, but for now, I will dig into it. Hopefully I will have a fix out by Monday at the latest.

**Comment by m50** [comment]
Hi, I created a beta release for the fix, if you could test it.

Just go into HACS, select Redownload, and enable Beta releases, and select `v1.0.5-beta`. This should fix the issues for everyone. Please let me know if everyone's issues are fixed, and I can merge and release a `v1.0.5` proper. I don't want to release a `v1.0.5` without knowing the issues are properly and finally fixes for everyone, as releasing code that I didn't personally have a hand in is how we got here.

I apologize for the length of time this took to fix. Now that I was able to debug the issue properly, it makes crystal clear sense what the issue was.

**Comment by Rafaille** [comment]
Works great now!
Thank you so much for the amazing support.

**Comment by mikey60** [comment]
Thanks for the quick support.  Unfortunately I will not be home until
Monday to test it.

Version 1.05 works great! Thank you!

**Comment by styphonthal** [comment]
The 1.0.5 fixed my problem.

**Comment by maxi1134** [comment]
1.0,5 Fixed #19 for me! Functional now with LocalAI

**Comment by m50** [comment]
I have merged the fix and released v1.0.5 officially now. Thanks for validating y'all, glad we could get this fixed :)
